lib/spy.js

```diff
@@ -13,6 +13,165 @@ module.exports = function (chai, _) {
   var Assertion = chai.Assertion
     , flag = _.flag
     , i = _.inspect
+    , ID_KEY = typeof Symbol === 'undefined' ? '__id' : Symbol('id')
+    , STATE_KEY = typeof Symbol === 'undefined' ? '__state' : Symbol('state')
+    , spyAmount = 0
+    , DEFAULT_SANDBOX = new Sandbox()
+
+  /**
+   * # Sandbox constructor (function)
+   *
+   * Initialize new Sandbox instance
+   *
+   * @returns new sandbox
+   * @api private
+   */
+
+  function Sandbox() {
+    this[STATE_KEY] = {};
+  }
+
+  /**
+   * # Sandbox.on (function)
+   *
+   * Wraps an object method into spy assigned to sandbox. All calls will
+   * pass through to the original function.
+   *
+   *      var spy = chai.spy.sandbox();
+   *      var isArray = spy.on(Array, 'isArray');
+   *
+   * @param {Object} object
+   * @param {String} method name to spy on
+   * @returns created spy
+   * @api public
+   */
+
+  Sandbox.prototype.on = function (object, methodName) {
+    var method = chai.spy('object.' + methodName, object[methodName]);
+
+    method[ID_KEY] = ++spyAmount;
+    this[STATE_KEY][method[ID_KEY]] = method;
+    method.__spy.tracked = {
+      object: object
+      , methodName: methodName
+      , originalMethod: object[methodName]
+      , isOwnMethod: object.hasOwnProperty(methodName)
+    };
+    object[methodName] = method;
+
+    return method;
+  };
+
+  /**
+   * # Sandbox.object (function)
+   *
+   * Creates an object with spied methods if first argument is a String.
+   * Wraps passed object's methods with spies if first argument is an Object.
+   * In case, if first parameter is an object and second is omit wraps
+   * all enumerable methods of passed object.
+   *
+   *      var spy = chai.spy.sandbox();
+   *      var object = spy.object('Array', [ 'push', 'pop' ]);
+   *      var array = spy.object([], [ 'push', 'pop' ]);
+   *
+   * @param {String|Object} name object or object name
+   * @param {String[]} [methods] method names
+   * @returns object with spied methods
+   * @api public
+   */
+
+  Sandbox.prototype.object = function (name, methods) {
+    var defs = {};
+    var sandbox = this;
+    var object;
+
+    if (name && typeof name === 'object') {
+      object = name;
+      methods = methods || Object.keys(object);
+      methods.forEach(function (methodName) {
+        var method = object[methodName];
+
+        if (typeof method === 'function' && !method.__spy) {
+          sandbox.on(object, methodName);
+        }
```

Comment:
I don't like that there is a silent failure path here. If the method is already a spy, then doing nothing is acceptable (because the intent from the developer is that they want a spy, and it is one); however if the method doesn't exist - we should either throw, or create an empty spy (my preferred option).